Test for convergence in Sum

maurogaravello opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fix issue #14463.


#### Brief description of what is fixed or changed

A test for convergence in `is_convergent()` method for `Sum` is added. Such a test aims to treat the series whose general term is a product between a bounded term `a_n` and a term `b_n`, such
that the sum of `b_n` is absolutely convergent.

Files changed:
   sympy/concrete/summations.py
   sympy/concrete/tests/test_sums_products.py


#### Other comments

`test_sums_products.py` includes `assert Sum(sin(n)/n**3, (n, 1, oo)).is_convergent() is S.true`.

maurogaravello committed:
Sum(sin(n) / n**3, (n, 1, oo)).is_convergent() raises NotImplementedError

Fix issue 14463.

A test for convergence in is_convergent() method is added. Such a test aims to treat the series
whose general term is a product between a bounded term a_n and a term b_n, such
that the sum of b_n is absolutely convergent.

Files changed:
   sympy/concrete/summations.py
   sympy/concrete/tests/test_sums_products.py

A test is also added.

normalhuman commented on sympy/concrete/summations.py:
```diff
@@ -518,8 +518,9 @@ def is_convergent(self):
+        ### ----- Dirichlet and bounded times convergent tests ----- ###
         if order.expr.is_Mul:
+            ### -------------- Dirichlet tests -------------- ###
             a_n, b_n = order.expr.args[0], order.expr.args[1]
```
This assumes that there are only two args to Mul. This isn't necessarily the case, for example
```
>>> O(sin(n)*log(n)/n**2, (n, oo)).expr.args
(n**(-2), log(n), sin(n))
```
As a result, `Sum(sin(n)*log(n)/n**2, (n, 1, oo)).is_convergent()` still fails to decide convergence. 

Even more troubling is the possibility that the args after the second affect the convergence of the series, which would invalidate the results of both Dirichlet test and bounded times convergent tests, as currently implemented. 

Two possible solutions: 

1. Restrict to 2-argument Mul 
2. Iterate through all possible partitions of Mul.args in two nonempty subsets, generating `a_n`, `b_n` for  the Dirichlet and bounded-convergent tests to try. This should be doable with `itertools.combinations`: say, there are N args, then loop 
```
args = order.expr.args
argset = set(args)
N = len(args)
for n in range(1, N):
    for a_tuple in itertools.combinations(args, n):
        b_set = argset - set(a_tuple)
        a_n = Mul(*a_tuple)
        b_n = Mul(*b_set)
        # call _dirichlet(a_n, b_n) and _bounded(a_n, b_n)
        # no need to try both orders of a_n and b_n, as, they both will be generated by itertools 
```

maurogaravello replied:
You right. This issue was also present in the old version of the Dirichlet test.

As regards the solution you proposed, I prefer the second one, because the first seems too restrictive. Do you prefer to write the code yourself, or will I do a commit to my PR?

maurogaravello committed:
Dirichlet and bounded times absolutely convergent test

Applying these tests, the general term could be a product of terms.
If the number of terms in the product is bigger than 2, then the previous
version of the tests could fail.

asmeurer commented:
You might put a TODO on the combinations code to improve the heuristics there. The code that is there now grows like 2^n with the number of args in the Mul, which, especially combined with the potentially slow checks inside the loop, could make this test extremely slow for larger summation expressions. 

Some thoughts:

- If the abs(Sum(b_n)) <= M test fails, then so will abs(Sum(b_n*y)) <= M
- Constants should be pulled out (or is that already done elsewhere?)
- It might be possible to try the combinations in a smarter order.

normalhuman commented:
The constants are pulled out by O. 
```
>>> O(3*log(n)/n**2, (n, oo)).expr.args
(n**(-2), log(n))
```

The order of terms returned by O seems alphabetical. 
```
>>> O(3*n*log(n)*exp(-n)*sin(n)*cos(n)*atan(n), (n, oo)).expr.args
(n, atan(n), cos(n), exp(-n), log(n), sin(n))
```

maurogaravello committed:
Merge remote-tracking branch 'upstream/master' into series_dirichlet

The conflit for
	sympy/concrete/tests/test_sums_products.py
was resolved.

Also a TODO was added in the code for explaining the heuristics of
the test for convergence.

maurogaravello committed:
Merge branch 'series_dirichlet' of https://github.com/maurogaravello/sympy into series_dirichlet

The conflict in the file
	sympy/concrete/tests/test_sums_products.py
is resolved.

A TODO for explaining the heuristics was added.

maurogaravello commented:
It seems that the merge conflict is resolved.

I put a TODO to explain the heuristics of the test.

normalhuman commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1011,12 +1009,15 @@ def test_issue_10156():
 
+def test_sin_times_absolutely_convergent():
+    assert Sum(sin(n) / n**3, (n, 1, oo)).is_convergent() is S.true
+    assert Sum(sin(n) * log(n) / n**3, (n, 1, oo)).is_convergent() is S.true
```
Strangely, `Sum(sin(n)*log(n)/n**3, (n, 2, oo)).is_convergent()` raises NotImplementedError. The only difference is using n=2 starting index instead of 1. What gives?

normalhuman replied:
This isn't a problem with new code here, though.
```
>>> Sum(log(n)/n**3, (n, 1, oo)).is_absolutely_convergent()
True
>>> Sum(log(n)/n**3, (n, 2, oo)).is_absolutely_convergent()
NotImplementedError
```
I think this can be left as a separate issue, as it's with another method; it just came up in testing here.

maurogaravello replied:
Does the failed test is related to this?

In my pc, the test `test_discrete_rv.py` works.

normalhuman replied:
The failed test is a random failure of a sampling method, which is random by nature. Another annoyance of having sampling methods. I restarted that test.

maurogaravello replied:
The issue `Sum(log(n)/n**3, (n, 1, oo)).is_absolutely_convergent()` is due to the integral test. If the series starts from `2`, then `maxima` in line 514 is an empty set.

maurogaravello replied:
Thanks for restarting the test.

normalhuman replied:
You are right about indentation. Would you like to correct that with this PR, or leave it for someone else?

maurogaravello replied:
I can leave it for someone else. It can be labelled as `easy to fix`.
However, it seems also necessary to add a flag, because sometimes `check_interval` is not assigned and some tests fail.
